File: src/features/remote-spy/types/remote-spy.ts
export type RemoteDirection = "outgoing" | "incoming";
export type RemoteClass =
  | "RemoteEvent"
  | "RemoteFunction"
  | "UnreliableRemoteEvent";

/**
 * Represents a single remote call event
 */
export interface RemoteCall {
  id: number;
  timestamp: Date;
  direction: RemoteDirection;
  arguments: RemoteArgument[];
  returnValue?: RemoteArgument;
  callingScriptName?: string;
  callingScriptPath?: string;
}

/**
 * Represents an argument or return value
 */
export interface RemoteArgument {
  type: string;
  value: string;
}

/**
 * Represents a unique remote (identified by numeric instance ID)
 */
export interface Remote {
  id: number; // Unique numeric ID for the remote instance
  name: string;
  path: string;
  class: RemoteClass;
  calls: RemoteCall[];
}

/**
 * Filter state for remote spy
 */
export interface RemoteSpyFilters {
  directions: RemoteDirection[];
  classes: RemoteClass[];
  search: string;
}

/**
 * Represents an attached client for remote spy
 */
export interface RemoteSpyClient {
  id: string;
  username: string;
}

File: src-tauri/src/models/remote_spy.rs
use serde::{Deserialize, Serialize};

/// Represents an argument or return value
#[derive(Debug, Serialize, Deserialize, Clone)]
pub struct RemoteArgument {
    #[serde(rename = "type")]
    pub arg_type: String,
    pub value: String,
}

/// Event payload for remote spy call
#[derive(Debug, Clone, Serialize)]
pub struct RemoteCallEvent {
    #[serde(rename = "callId")]
    pub call_id: u32,
    #[serde(rename = "remoteId")]
    pub remote_id: u32,
    pub name: String,
    pub path: String,
    pub class: String,
    pub direction: String,
    pub timestamp: String,
    pub arguments: Vec<RemoteArgument>,
    #[serde(skip_serializing_if = "Option::is_none")]
    #[serde(rename = "returnValues")]
    pub return_values: Option<Vec<RemoteArgument>>,
    #[serde(skip_serializing_if = "Option::is_none")]
    #[serde(rename = "callingScriptName")]
    pub calling_script_name: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    #[serde(rename = "callingScriptPath")]
    pub calling_script_path: Option<String>,
}
